addons/mail/models/discuss/discuss_channel.py

```diff
@@ -673,6 +673,80 @@ def _add_members(
+        """
+        Send channel invitation emails to a list of email addresses.
+
+        - Existing members' email addresses are ignored.
+        - An invitation can only be re-sent to the same email address for the
+        same channel after 5 minutes.
+
+        :param addresses: List of email addresses to invite.
+        :type addresses: list[str]
+
+        """
+        self.check_access("read")
+        if not self._allow_invite_by_email():
+            raise UserError(
+                self.env._("Inviting by email is not allowed for this channel type (%s).")
+                % self.channel_type
+            )
+        addresses = {
+            norm for addr in addresses if addr and (norm := email_normalize(addr))
+        }
+        # Removing already sent email in the past 5 minutes to avoid spamming. However, it's
+        # sometimes acceptable to re-send an invitation (e.g. the recipient deleted the email).
+        #
+        # sudo - mail.mail: accessing mail records to determine which emails were already sent
+        # is acceptable.
+        addresses -= set(
+            self.env["mail.mail"]
+            .sudo()
+            .search_fetch(
+                [
+                    ("references", "=", f"channel_invite_{self.id}"),
+                    ("create_date", ">", "-5M"),
+                    ("email_to", "in", addresses),
+                ],
+                ["email_to"],
+            )
+            .mapped("email_to"),
+        )
+        # Removing emails linked to members of this channel.
+        member_domain = Domain([("channel_id", "=", self.id)])
+        member_domain &= Domain([("guest_id.email", "in", addresses)]) | Domain(
+            [("partner_id.email", "in", addresses)]
+        )
+        addresses -= set(
+            self.env["discuss.channel.member"]
+            .search_fetch(member_domain, ["partner_id", "guest_id"])
+            .mapped(lambda m: m.partner_id.email or m.guest_id.email)
+        )
+        to_create = []
+        for addr in addresses:
+            body = self.env["ir.qweb"]._render(
+                "mail.discuss_channel_invitation_template",
+                {
+                    "channel": self,
+                    "user": self.env.user,
+                    "base_url": self.env["ir.config_parameter"].get_base_url(),
+                    "email_to": addr,
+                },
+                minimal_qcontext=True,
+            )
+            to_create.append(
+                {
+                    "subject": self.env._("%(author_name)s has invited you to a channel")
+                    % {"author_name": self.env.user.name},
+                    "email_from": self.env.user.partner_id.email_formatted,
+                    "references": f"channel_invite_{self.id}",
+                    "email_to": addr,
+                    "body_html": body,
+                },
+            )
+        if to_create:
+            self.env["mail.mail"].sudo().create(to_create).send()
```

Comment:
Never send in a given transaction, unless strong reasons. Indeed we don't want to force people to wait / have issues due to email server errors. Use the cron (aka don't send, just set as outgoing).